### DISCUSSION TITLE
How can I dynamically modify the AST in Pyright to update variable type information?

### DISCUSSION THREAD

**Discussion by jianzs** [microsoft/pyright#8174] [Q&A]
I am seeking assistance on integrating a [particular feature](https://github.com/pluto-lang/pluto/issues/240) into a development tool that leverages Pyright as its static analysis engine. 

My objective is to evaluate local variables within a function based on the arguments passed to that function during static analysis. 

To leverage Pyright’s type inference capabilities, I hope to add an assignment statement for a variable directly in the AST of a function body. This would effectively define the function's parameters as local variables within that function, allowing Pyright to infer the types of subsequent local variable usages, and by extension, deduce their literal values when called.

Consider the following Python code as an example:

```python
def foo(arg: str):
    # Placeholder for dynamically added assignment: arg = "hello" / "world"
    processed = "arg-" + arg
    bar(processed)

foo("hello")
foo("world")
```

My goal is to dynamically insert assignments, such as `arg = "hello"` or `arg = "world"`, into the `foo` function body for each invocation of `foo`. Following this insertion, I expect Pyright to infer the type of `processed` as either "arg-hello" or "arg-world", based on the modified AST. This, in turn, should enable Pyright to deduce that the argument passed to `bar` during these calls is of type "arg-hello" or "arg-world".

Is there a method to programmatically add a statement node to the AST used by Pyright, and subsequently refresh or re-analyze the AST to update the type information for variables within the AST accordingly?

**Comment by erictraut** [comment]
It would be a pretty significant modification to support this, as it breaks some fundamental assumptions in pyright's design. If a symbol has a declared type (as `arg` does in your example above), pyright honors that declaration and doesn't allow redefinitions. Plus, the parse tree is assumed to be static once it is created. Any modification that occurs (e.g. due to new user input) causes the entire parse tree to be discarded and regenerated from scratch.

**Comment by jianzs** [comment]
@erictraut Thank you for responding! Do you have any ideas on how to determine the value of a local variable? Any concepts, tools, or thoughts are welcome.

Currently, I am considering constructing a tree recursively based on the declaration of the local variable. The parameter node will serve as a placeholder on the tree, and input the actual argument during function invocation to the corresponding placeholder. Finally, I would execute the tree to get the value of the local variable. However, I am not sure if this is the best way to solve this problem.
